posts/inside-rust/2020-05-25-new-inline-asm.md

```diff
@@ -0,0 +1,130 @@
+  implemented the new `asm!` syntax in nightly.
+
+Here's an example of using the new inline assembly syntax, to print a message
+to standard output using a direct [`write`
+syscall](https://man7.org/linux/man-pages/man2/write.2.html) on x86-64 Linux:
+
+```rust
+#![feature(asm)]
+
+fn main() {
+    let buf = "Hello from asm!\n";
+    let ret: i32;
+    unsafe {
+        asm!(
+            "syscall",
+            in("rax") 1, // syscall number
+            in("rdi") 1, // fd
+            in("rsi") buf.as_ptr(),
+            in("rdx") buf.len(),
+            out("rcx") _, // clobbered by syscalls
+            out("r11") _, // clobbered by syscalls
+            lateout("rax") ret,
+        );
+    }
+    println!("write returned: {}", ret);
+}
+```
+
+(You can [try this example on the
+playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=e983a5f5cffa51f4320f1176465d3a56).)
+
+The example above specifies the exact inputs, outputs, and clobbers required by
+the Linux syscall calling convention. You can also provide inputs and outputs
+via arbitrary registers, and the compiler will select appropriate registers for
+you. The following example uses [bit manipulation
+instructions](https://en.wikipedia.org/wiki/Bit_Manipulation_Instruction_Sets)
+to compute the bit numbers of all set bits in a value, and stores them in a
+slice of memory:
+
+```rust
+#![feature(asm)]
+
+fn main() {
+    let mut bits = [0u8; 64];
+    for value in 0..=1024u64 {
+        let popcnt;
+        unsafe {
+            asm!("
+                popcnt {popcnt}, {v}
+                2:
+                blsi rax, {v}
+                jz 1f
+                xor {v}, rax
+                tzcnt rax, rax
+                stosb
+                jmp 2b
+                1:
+                ",
+                v = inout(reg) value => _,
+                popcnt = out(reg) popcnt,
+                out("rax") _, // scratch
+                inout("rdi") bits.as_mut_ptr() => _,
+            );
+        }
+        println!("bits of {}: {:?}", value, &bits[0..popcnt]);
+    }
+}
+```
+
+(You can [try this example on the
+playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=38874735e48aa20289f23f5a3cbeae0c).
+Note that this code serves to demonstrate inline assembly, not to demonstrate
+an efficient implementation of any particular algorithm.)
+
+Notice that `value` and `popcnt` have registers selected for them, while
+`bits.as_mut_ptr()` must go in the `rdi` register for use with the `stosb`
+instruction.
+
+For full details on the new `asm!` syntax, see [RFC
+2873](https://github.com/Amanieu/rfcs/blob/inline-asm/text/0000-inline-asm.md).
+Please try it out, and [report any bugs via the rust issue
+tracker](https://github.com/rust-lang/rust/issues/) with the tag `F-asm`.  You
+can also [report your experiences on the tracking
+issue](https://github.com/rust-lang/rust/issues/72016).
```

Comment:
Should we perhaps instead direct to an internals or Zulip thread? I expect that the tracking issue is otherwise likely to relatively quickly get lots of comments, and "protecting" it seems worthwhile.

Reply:
I think I'll just drop the reference to the tracking issue entirely. People can already find it via the other github links if they're dedicated. We should also have a note at the top of the tracking issue saying something like "If you came here to report a bug in inline assembly, please report it as a separate issue and link to this one; don't post a comment here".